src/shogun/mathematics/linalg/internal/implementation/Add.h

```diff
@@ -55,20 +55,89 @@ namespace linalg
 namespace implementation
 {
 
-/** Generic class which is specialized for different backends to perform addition */
+/**
+ * @brief Generic class which is specialized for different backends to perform addition
+ */
 template <enum Backend, class Matrix>
 struct add
 {
 	/** Scalar type */
 	typedef typename Matrix::Scalar T;
 	
-	/** Performs the operation C = alpha*A + beta*B. Works for both matrices and vectors */
+	/**
+	 * Performs the operation C = alpha*A + beta*B. Works for both matrices and vectors
+	 * @param A first matrix
+	 * @param B second matrix
+	 * @param C matrix to store the result
+	 * @param alpha constant to be multiplied by the first matrix
+	 * @param beta constant to be multiplied by the second matrix
+	 */
 	static void compute(Matrix A, Matrix B, Matrix C, T alpha, T beta);
 };
 
+/**
+ * @brief Specialization of add for the Native backend
+ */
+template <> template <class Matrix>
+struct add<Backend::NATIVE, Matrix>
+{
+	typedef typename Matrix::Scalar T;
+
+	/**
+	 * Performs the operation C = alpha*A + beta*B.
+	 * @param A first matrix
+	 * @param B second matrix
+	 * @param C matrix to store the result
+	 * @param alpha constant to be multiplied by the first matrix
+	 * @param beta constant to be multiplied by the second matrix
+	 */
+	static void compute(SGMatrix<T> A, SGMatrix<T> B, SGMatrix<T> C,
+		T alpha, T beta)
+	{
+		REQUIRE(A.num_rows == B.num_rows && B.num_rows == C.num_rows,
+			"Matrices should have same number of rows!\n");
+		REQUIRE(A.num_cols == B.num_cols && B.num_cols == C.num_cols,
+			"Matrices should have same number of columns!\n");
+		compute(A.matrix, B.matrix, C.matrix, alpha, beta, A.num_rows*A.num_cols);
+	}
+	
+	/**
+	 * Performs the operation C = alpha*A + beta*B.
+	 * @param A first vector
+	 * @param B second vector
+	 * @param C vector to store the result
+	 * @param alpha constant to be multiplied by the first vector
+	 * @param beta constant to be multiplied by the second vector
+	 */
+	static void compute(SGVector<T> A, SGVector<T> B, SGVector<T> C,
+		T alpha, T beta)
+	{
+		REQUIRE(A.vlen == B.vlen, "Vectors should have same length!\n");
+		compute(A.vector, B.vector, C.vector, alpha, beta, A.vlen);
+	}
+
+	/**
+	 * Performs the operation C = alpha*A + beta*B. Vectors or Matrices passed as pointers
+	 * @param A first vector
+	 * @param B second vector
+	 * @param C vector to store the result
+	 * @param alpha constant to be multiplied by the first vector
+	 * @param beta constant to be multiplied by the second vector
+	 * @param len length of the vectors/matrices
+	 */
+	static void compute(T* A, T* B, T* C,
+		T alpha, T beta, index_t len)
+	{
+		for (int32_t i=0; i<len; i++)
```

Comment:
Also what all operations would we generally like to do in parallel? Do we always use openmp or something else as well? Just so that I take care of it in future :)